ui/src/containers/appView/AppPageContainer.js

```diff
@@ -36,36 +43,43 @@ class AppPageContainer extends React.Component {
     this.unblockHistory();
   }
 
+  /**
+   * Handles when user clicks Confirm in the Save App Model.
+   * This function closes the model, then it saves the changed versions to the server.
+   * When that is complete, it fetches fresh app data from the server.
+   *
+   * @memberof AppPageContainer
+   */
   onConfirmSaveApp () {
+    const { app: { id, deployedVersions: currentVersions }, savedData: { deployedVersions: savedVersions } } = this.props;
+
     this.props.toggleSaveAppModal();
-    // TODO: Make a PUT request to API
-    // to update the App versions
-  }
 
-  isAppVersionsDirty () {
-    const currentVersions = this.props.app.deployedVersions;
-    const savedVersions = this.props.savedData.deployedVersions;
+    const dirtyVersions = AppService.getDirtyVersions(savedVersions, currentVersions);
 
-    var isDirty = false;
+    this.props.saveAppVersions(id, dirtyVersions);
+  }
 
-    for (var i = 0; i < currentVersions.length; i++) {
-      var currentVersionValues = Object.values(currentVersions[ i ]).toString();
-      var savedVersionValues = Object.values(savedVersions[ i ]).toString();
+  /**
+   * Checks if the app versions form is dirty.
+   *
+   * @returns {Boolean} is the form dirty
+   * @memberof AppPageContainer
+   */
+  isAppVersionsDirty () {
+    const { app: { deployedVersions: currentVersions },
+      savedData: { deployedVersions: savedVersions } } = this.props;
 
-      if (currentVersionValues !== savedVersionValues) {
-        isDirty = true;
-        break;
-      }
-    }
+    const dirtyItems = AppService.getDirtyVersions(savedVersions, currentVersions);
 
-    return isDirty;
+    return !!dirtyItems.length;
```

Comment:
I don't believe I have ever seen this syntax before `!!`. Do you have a link to this?

Reply:
It is just a double `!` operator - not not. The opposite of `!dirtyItems.length`.

https://stackoverflow.com/a/784946

Reply:
Perfect, thanks